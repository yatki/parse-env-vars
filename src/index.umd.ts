/* istanbul ignore file */
import { readEnv } from '.';

export default readEnv;
