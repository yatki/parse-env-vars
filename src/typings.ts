type FormatFunction = (str: string) => string;

interface Source {
  [key: string]: string | undefined;
}

interface SanitizeOptions {
  object?: boolean;
  array?: boolean;
  bool?: boolean;
  int?: boolean;
  float?: boolean;
}

type Format = 'camelcase' | 'pascalcase' | 'lowercase' | 'uppercase';

interface ReadEnvOptions {
  source: Source;
  separator: string;
  includePrefix: boolean;
  format: boolean | Format | FormatFunction;
  sanitize: boolean | SanitizeOptions;
  sort: boolean;
}

type ReadEnvResult = Record<string, any>;

export {
  ReadEnvOptions,
  Format,
  SanitizeOptions,
  FormatFunction,
  ReadEnvResult,
};
