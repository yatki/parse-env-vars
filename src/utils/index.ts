export * from './options';
export * from './format';
export * from './sanitize';

const trimLeft = (str: string, charList: string): string =>
  str.replace(new RegExp(`^[${charList}]+`), '');

export { trimLeft };
