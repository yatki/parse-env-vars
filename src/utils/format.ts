import camelcase from 'camelcase';
import { FormatFunction } from '../typings';

const formatters: Record<string, FormatFunction> = {
  camelcase: (str: string): string => camelcase(str.toLowerCase()),
  pascalcase: (str: string): string =>
    camelcase(str.toLowerCase(), { pascalCase: true }),
  lowercase: (str: string): string => str.toLowerCase(),
  uppercase: (str: string): string => str.toUpperCase(),
};

const formatKey = (
  key: string,
  format: boolean | string | FormatFunction,
): string => {
  if (typeof format === 'string') {
    return formatters[format](key);
  }

  if (typeof format === 'function') {
    return format(key);
  }

  return key;
};

export { formatKey };
