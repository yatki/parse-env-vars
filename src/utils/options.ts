import { ReadEnvOptions, SanitizeOptions } from '../typings';

const defaultOptions: ReadEnvOptions = {
  source: process.env,
  format: 'camelcase',
  separator: '__',
  sort: true,
  sanitize: {
    object: true,
    array: true,
    bool: true,
    int: true,
    float: true,
  },
  includePrefix: false,
};

const getOptions = (userOptions: Partial<ReadEnvOptions>): ReadEnvOptions => {
  if (typeof userOptions.sanitize === 'object') {
    // eslint-disable-next-line no-param-reassign
    userOptions.sanitize = {
      ...(defaultOptions.sanitize as SanitizeOptions),
      ...userOptions.sanitize,
    };
  }
  return { ...defaultOptions, ...userOptions };
};

export { defaultOptions, getOptions };
