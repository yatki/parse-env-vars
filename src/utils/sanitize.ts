import { SanitizeOptions } from '../typings';

const isObject = (value: string): boolean =>
  value[0] === '{' && value[value.length - 1] === '}';

const isArray = (value: string): boolean =>
  value[0] === '[' && value[value.length - 1] === ']';

const isInt = (value: string): boolean => /^-?\d+$/.test(value);

const isFloat = (value: string): boolean => /^-?\d*(\.\d+)$/.test(value);
const isBool = (value: string): boolean => {
  const boolValue = value.toLowerCase();
  return boolValue === 'true' || boolValue === 'false';
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const sanitize = (value: string, options: SanitizeOptions): any => {
  // Sanitize Object
  if (options.object && isObject(value)) {
    try {
      return JSON.parse(value);
    } catch (err) {
      return value;
    }
  }

  // Sanitize Array
  if (options.array && isArray(value)) {
    try {
      return JSON.parse(value);
    } catch (err) {
      return value;
    }
  }

  // Sanitize Integer
  if (options.int && isInt(value)) {
    return parseInt(value, 10);
  }

  // Sanitize Float
  if (options.float && isFloat(value)) {
    return parseFloat(value);
  }

  // Sanitize Boolean
  if (options.bool && isBool(value)) {
    return value.toLowerCase() === 'true';
  }

  // Return raw value
  return value;
};

export { sanitize };
