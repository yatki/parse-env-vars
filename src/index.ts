import { ReadEnvOptions, ReadEnvResult } from './typings';
import { getOptions, sanitize, formatKey, trimLeft } from './utils';

const readEnv = (
  prefix?: string,
  options: Partial<ReadEnvOptions> = {},
): ReadEnvResult => {
  // Options
  const {
    format,
    source = process.env,
    includePrefix,
    sanitize: sanitizeOptions,
    separator,
  } = getOptions(options);

  let keys = Object.keys(source);

  // Filter keys by prefix
  if (typeof prefix === 'string' && prefix) {
    keys = keys.filter((value) => value.indexOf(prefix) === 0);
  }

  // Process keys
  return keys.reduce((result, key) => {
    let envValue = source[key]?.trim();
    let envKey = key;

    // Sanitize
    if (typeof sanitizeOptions === 'object' && envValue) {
      envValue = sanitize(envValue, sanitizeOptions);
    }

    // Remove Prefix
    if (!includePrefix && typeof prefix === 'string' && prefix) {
      envKey = envKey.replace(prefix, '');
    }

    // Trim left underscore
    envKey = trimLeft(envKey, '_');

    let deepKeys = [envKey];
    // Process deep object
    if (typeof separator === 'string' && separator) {
      deepKeys = envKey.split(separator);
    }

    deepKeys.reduce((acc, item, index) => {
      const deepKey = formatKey(item, format);
      if (index === deepKeys.length - 1) {
        acc[deepKey] = envValue;
      } else if (
        typeof acc[deepKey] !== 'object' ||
        Array.isArray(acc[deepKey])
      ) {
        acc[deepKey] = {};
      }
      return acc[deepKey];
    }, result);

    // and DONE!
    return result;
  }, {} as ReadEnvResult);
};

export { readEnv };
