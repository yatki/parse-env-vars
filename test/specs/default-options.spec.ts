import { defaultOptions } from '../../src/utils';

it('should match the snapshot', () => {
  expect(defaultOptions).toMatchSnapshot({
    source: expect.any(Object),
  });
});
