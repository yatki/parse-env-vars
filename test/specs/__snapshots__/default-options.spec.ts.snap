// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
Object {
  "format": "camelcase",
  "includePrefix": false,
  "sanitize": Object {
    "array": true,
    "bool": true,
    "float": true,
    "int": true,
    "object": true,
  },
  "separator": "__",
  "source": Any<Object>,
}
`;
